File: leetcode-105.py
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        m = len(preorder)
        n = len(inorder)
        dic = dict()
        for i in range(n):
            dic[inorder[i]] = i
        def dfs(l=0,r=m-1, ll=0, rr=n-1):
            if ll > rr or l > r:
                return None
            v = preorder[l]
            u = TreeNode(v)
            k = dic[v]-ll
            u.left = dfs(l+1, l+k, ll, dic[v]-1)
            u.right = dfs(l+k+1, r, dic[v]+1, rr)
            return u
        
        return dfs()


if __name__ == "__main__":
    so = Solution()
    print(so.buildTree([3,9,20,15,7], [9,3,15,20,7]))

File: leetcode/leetcode-105.cpp
/*
 *@File Name          :leetcode-105.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/26 21:36
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "vector"
#include "algorithm"
#include "map"

using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    map<int, int> ma;
    int k;

    TreeNode *dfs(vector<int> &preorder, vector<int> &inorder, int l, int r) {
        if (l > r) return nullptr;
        TreeNode *t = new TreeNode(preorder[k]);
        int mid = ma[preorder[k++]];
        t->left = dfs(preorder, inorder, l, mid - 1);
        t->right = dfs(preorder, inorder, mid + 1, r);
        return t;
    }

    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        for (int i = 0; i < inorder.size(); i++)
            ma[inorder[i]] = i;
        k = 0;
        return dfs(preorder, inorder, 0, preorder.size() - 1);
    }
};